### PR TITLE
Installers Add riscv64 support

### PR DESCRIPTION
**Pull request by steelhead31**
Implement code changes to add riscv support for the installers packaging process (JDK21 onwards). Because the jfrog-cli tools are not available for riscv at this point, the installers process builds the packages on a riscv machine, before switching to an x64 node to push the files to artifactory.

### CODE REVIEW

**Comment by steelhead31** [comment]
This PR will also allow the installers process to be run for the as yet unpublished JDK21/riscv installers.

**Comment by steelhead31** [comment]
Test runs in various states can be seen here: [https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/](https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/) , but in particular, this one [https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/110/](https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/110/) has the riscv packages for debian , this one [https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/111/](https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/111/) for SUSE, and this one for RHEL : https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/112/

File: linux/Jenkinsfile [hunk 1]
```diff
@@ -174,18 +175,67 @@ def CheckAndUpload(String Target, String Distro, String BuildArch, String Versio
...
            case "Debian":
-               jf 'rt u **/build/ospackage/temurin-*${BUILDARCH}.deb deb/pool/main/t/temurin-${VERSION}/ --target-props=${DISTROLIST}deb.component=main;deb.architecture=${BUILDARCH} --flat=true'
-               break
+               // Not RISC Use JF CLI
+               if (BuildArch != "riscv64") {
+                 jf 'rt u **/build/ospackage/temurin-*${BUILDARCH}.deb deb/pool/main/t/temurin-${VERSION}/ --target-props=${DISTROLIST}deb.component=main;deb.architecture=${BUILDARCH} --flat=true'
+               }
+               if (BuildArch == "riscv64") {
+                     // RISC Use RT upload
+                     node("dockerBuild&&linux&&x64") {
```
Comment (karianna) [inline linux/Jenkinsfile:184]: Can we extract this duplicated code somehow?

File: linux/Jenkinsfile [hunk 2]
```diff
@@ -224,11 +278,15 @@ def jenkinsStepDeb() {
...
         }
-
+        if ("${DebARCH}" == 'riscv64') {
+            debLabel = 'riscv64&&dockerInstaller'
+        }
         // reallocate jenkins agent per element in list
         node("linux&&${debLabel}") {
             setup('Debian', "${DebARCH}")
-            tool name: 'jfrog-cli', type: 'jfrog'
+            if ("${DebARCH}" != 'riscv64') {
```
Comment (karianna) [inline linux/Jenkinsfile:287]: Is this logic correct, other logic above has riscv with the jfrog cli

Comment (steelhead31) [reply linux/Jenkinsfile:287]: Yes, the new way this works is to do the package build on a risc-v machine, publish those packages to jenkins, and then upload them from an x64 machine, as the jfrog-cli currently does not support riscv, so its a change to the process for riscv.. that said, if jfrog ever add riscv support to the jfrog cli, we'll be able to realign this functionality.

File: linux/Jenkinsfile (hunk 1 above)
Comment (steelhead31) [reply linux/Jenkinsfile]: Yup, im just testing a change to improve this.

**Comment by karianna** [comment]
Not sure if failures unrelated, I suspect so

**Comment by steelhead31** [comment]
> Not sure if failures unrelated, I suspect so

@karianna the failures are related to the fact we didn't publish s390x & riscv on the JDK22 release in March. I've added the code for these in preparation for the upcoming release.

**Review by tellison — APPROVED** [review]

**Comment by tellison** [comment]
Some significant changes in there - will need careful watching when running the release. Thanks for the update.

**Comment by steelhead31** [comment]
> Some significant changes in there - will need careful watching when running the release. Thanks for the update.

@tellison Yup, I've done a LOT of testing.. in this jenkins pipeline ( may not be visible ) without the upload bits enabled .. [https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/](https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/)    I've even tested a couple of things with the upload... I'll likely take responsibility for doing the installers in this next release, so I can keep an eye on this! :)

**Review by sxa — CHANGES_REQUESTED** [review]
I'm marking this as request changes only because I'd like to understand the reason for the Linux/s390x 22+36 inclusion - the name changes are not critical and given you've done a load of testing I don't want to force that to be redone for this.

I'd echo Tim's comment that this is a bit "big and scary" although we can test at least the RISC-V case on the current release to give us a little extra confidence before going into next week to mitigate that somewhat.

File: linux/jdk/debian/src/main/packaging/temurin/22/debian/rules [hunk 3]
```diff
@@ -3,13 +3,16 @@
...
 priority = 2211
 jvm_tools = jar jarsigner java javac javadoc javap jcmd jconsole jdb jdeprscan jdeps jfr jhsdb jimage jinfo jlink jmap jmod jpackage jps jrunscript jshell jstack jstat jstatd jwebserver keytool rmiregistry serialver jexec jspawnhelper
-
 amd64_tarball_url = https://github.com/adoptium/temurin22-binaries/releases/download/jdk-22%2B36/OpenJDK22U-jdk_x64_linux_hotspot_22_36.tar.gz
 amd64_checksum = bc3d99e816d0c373f424cd7aa2b6d3e8081a7189fe55c1561616922200ec8e47
 arm64_tarball_url = https://github.com/adoptium/temurin22-binaries/releases/download/jdk-22%2B36/OpenJDK22U-jdk_aarch64_linux_hotspot_22_36.tar.gz
 arm64_checksum = 4b52670caea44848cee893e35c804380817b6eff166cf64ee70ca2bfaac3d1c7
 ppc64el_tarball_url = https://github.com/adoptium/temurin22-binaries/releases/download/jdk-22%2B36/OpenJDK22U-jdk_ppc64le_linux_hotspot_22_36.tar.gz
 ppc64el_checksum = 8c062e934d95c639f97b4e51b968eed694a6653248727c3db8bc5e0e55cfd7f4
+s390x_tarball_url = https://github.com/adoptium/temurin22-binaries/releases/download/jdk-22.0.0%2B0/OpenJDK22U-jdk_s390x_linux_hotspot_22_36.tar.gz
```
[the remark this answers is not in the document]
Comment (steelhead31) [reply linux/jdk/debian/src/main/packaging/temurin/22/debian/rules:12]: Yup, its a surprisingly big job at the moment to add & remove platforms, and most of this PR was written in anticipation of the previous release having all platforms...

File: linux/Jenkinsfile [hunk 4]
```diff
@@ -193,6 +210,36 @@ def CheckAndUpload(String Target, String Distro, String BuildArch, String Versio
   }
 }
 
+// Riscv Upload Function
+def handleRISCUpload(String Distro) {
```
[the remark this answers is not in the document]
Comment (steelhead31) [reply linux/Jenkinsfile]: A relatively small change to the function name, and the 3 calls.. I've done it for consistency.

**Comment by steelhead31** [comment]
A full run of all packages for all architectures JDK21 has completed OK ( without uploads ).. https://ci.adoptium.net/job/doptium-packages-linux-pipeline_new-scottfryer/146/

**Pull request merged**